docs/DifferentiableTypes.md

```diff
@@ -0,0 +1,289 @@
+
+Richard Wei, Dan Zheng
+
+March 2019
+
+## Preface
+
+Mathematically speaking, only functions are "differentiable": only functions have derivatives and can be differentiated. In this document, the terminology "differentiable types" is used as a shorthand for "types that can be used as arguments and results of differentiable functions". This notion is important because not all types are "differentiable" in this sense. For example, types representing real numbers and vector spaces are "differentiable", but strings and integers are not.
+
+## Introduction
+
+Elementary calculus defines differentiation on real numbers: most people are familiar with this definition of "differentiation". However, differentiation is defined for many concepts across different branches of mathematics:
+* Scalar differentiation: differentiation on real numbers. This is taught in introductory calculus.
+* [Vector calculus]: a branch of mathematics that involves differentiation of vector fields.
+* [Differential geometry]: a branch of mathematics that involves differentiation of functions over manifolds.
+
+In Swift, we want to build a general system for differentiation that can represent all of these cases. Differentiation should not be limited to functions over specific types (e.g. functions over floating-point numbers); it should also work with functions whose parameters/result are custom types.
+
+This raises some questions: what kind of types can be used as arguments and results of differentiable functions, and how can we generalize them using a protocol?
+
+## Design overview
+
+The `Differentiable` protocol generalizes all types that can be used as arguments and results of differentiable functions.
+
+The compiler can automatically provide default implementations of `Differentiable` protocol requirements for struct types whose stored properties all conform to `Differentiable`.
+
+Here are some examples:
+
+```swift
+struct Vector: Differentiable, VectorNumeric {
+    // The compiler synthesizes all `Differentiable` protocol requirements
+    // when all stored properties conform to `Differentiable`.
+    var x, y, z: Float
+}
+
+// Differential operators like `gradient(at:in:)` just work!
+let v = Vector(x: 1, y: 2, z: 3)
+let 𝛁v = gradient(at: v) { v in (v + v).x }
+
+print(𝛁v)
+// Vector(x: 2.0, y: 0.0, z: 0.0)
+```
+
+A `Differentiable`-conforming type may have stored properties that are not meant to have a derivative with respect to `self`. Use the `@noDerivative` attribute to mark those properties; they will not have a corresponding entry in the synthesized `TangentVector`, `CotangentVector`, and `AllDifferentiableVariables` struct types.
+
+Here’s an example deep learning layer with some `@noDerivative` properties:
+
+```swift
+struct DenseLayer : Differentiable {
+    // These properties should have derivative values.
+    var weight: Tensor<Float>
+    var bias: Tensor<Float>
+
+    // These auxiliary properties should not have derivative values.
```

Comment:
It seems that these comments are line-wrapped based on the 80-column limit, not 100-column.